doc/rewriting-services.md

```diff
@@ -0,0 +1,204 @@
+downloads can go in one badge, and star rating vs point rating vs rating count
+can go in one badge, and same with various kinds of detail about a pull request.
+The hard limit (as of now anyway) is *one category per service*.
+
+3. If the tests haven’t been split up, split them up too and make sure they
+still pass.
+
+## Get the route working
+
+1. Disable the legacy service by adding a `return` at the top of
+`registerLegacyRouteHandler()`.
+
+2. Set up the route for one of the badges. First determine if you can express
+the route using a `pattern`. A `pattern` is the simplest way to declare the
+route, also the most readable, and will be useful for displaying a badge builder
+with fields in the front end and generating badge URLs programmatically. One
+limitation to keep in mind is that, at present, the trailing parameter of a
+pattern can't be optional. If the last part of a route is optional, like a
+branch, you will need to use a `format` regex.
+
+3. When creating the initial route, you can stub out the service. A minimal
+service extends BaseJsonService (or BaseService, or one of the others), and
+defines `route()` and `handle()`. `defaultBadgeData` is optional but suggested:
+
+```js
+const BaseJsonService = require('../base-json')
+
+class ExampleDownloads extends BaseJsonService {
+  static get route() {
+    return {
+       base: 'example/d',
+       pattern: ':param1/:param2’:',
+     }
+   }
+
+   static defaultBadgeData() {
+     return { label: 'downloads' } // or whatever
+   }
+
+   async handle({ param1, param2 }) {
+     return { message: 'hello' }
+   }
+}
+```
+
+4. We don’t have really good tools for debugging matches, so the best you can do
+is run a subset of your tests. To run a single service test, add `.only()`
+somewhere in the chain, and run `npm run test:services:trace — —only=example`.
+
+```js
+t.create('build status')
+  .get('/pip.json')
+  .only() // Prevent this ServiceTester from running its other tests.
+  .expectJSONTypes(
+    Joi.object().keys({
+      name: 'docs',
+      value: Joi.alternatives().try(isBuildStatus, Joi.equal('unknown')),
+    })
+  )
+```
+
+5. Presumably the test will fail, though by examining the copious output, you
+can confirm the route was matched and the named parameters mapped successfully.
+Since you'll have just run the tests on the old code (right?) you'll know you
+haven't inadvertently changed the route (an easy mistake to make).
+
+6. If the legacy service had a base URL and you've changed it, you’ll need to
+update the tests _and_ the examples. Take care to do that.
+
+## Implement `render()` and `handle()`
+
+1. Once the route is working, fill out `render()` and `handle()`. If there’s a
+single service, you can implement fetch as a method or a function at the top of
+the file. If there are more than one service which share fetching code, you can
+put the fetch function in `example-common.js` like this one for github:
```

Comment:
Lets try and keep the terminology consistent. Earlier we were using `myservice-version.service.js` but now we're using `example-common.js`. Lets use `example` or `myservice` as a prefix everywhere. Given the comment about overloading the term 'service', maybe using `example` is better?